pvlib/iotools/psm3.py

```diff
@@ -154,30 +141,132 @@ def get_psm3(latitude, longitude, api_key, email, names='tmy', interval=60,
-    # get the column names so we can set the dtypes
-    columns = fbuf.readline().split(',')
-    columns[-1] = columns[-1].strip()  # strip trailing newline
-    dtypes = dict.fromkeys(columns, float)  # all floats except datevec
-    dtypes.update(Year=int, Month=int, Day=int, Hour=int, Minute=int)
-    data = pd.read_csv(
-        fbuf, header=None, names=columns, dtype=dtypes,
-        delimiter=',', lineterminator='\n')  # skip carriage returns \r
-    # the response 1st 5 columns are a date vector, convert to datetime
-    dtidx = pd.to_datetime(
-        data[['Year', 'Month', 'Day', 'Hour', 'Minute']])
-    # in USA all timezones are intergers
-    tz = 'Etc/GMT%+d' % -header['Time Zone']
-    data.index = pd.DatetimeIndex(dtidx).tz_localize(tz)
+    return read_psm3(fbuf)
+
+
+def read_psm3(filename):
+    """
+    Read an NSRDB [1]_ PSM3 weather file (formatted as SAM CSV [2]_).
+
+    Parameters
+    ----------
+    filename: string or file-like object
+        Filename or file-like object of data to read.
+
+    Returns
+    -------
+    headers : dict
+        metadata from NREL PSM3 about the record, see notes for fields
+    data : pandas.DataFrame
+        timeseries data from NREL PSM3
+
+    Notes
+    -----
+    The return is a tuple with two items. The first item is a header with
+    metadata from NREL PSM3 about the record containing the following fields:
+
+    * Source
+    * Location ID
+    * City
+    * State
+    * Country
+    * Latitude
+    * Longitude
+    * Time Zone
+    * Elevation
+    * Local Time Zone
+    * Clearsky DHI Units
+    * Clearsky DNI Units
+    * Clearsky GHI Units
+    * Dew Point Units
+    * DHI Units
+    * DNI Units
+    * GHI Units
+    * Solar Zenith Angle Units
+    * Temperature Units
+    * Pressure Units
+    * Relative Humidity Units
+    * Precipitable Water Units
+    * Wind Direction Units
+    * Wind Speed
+    * Cloud Type -15
+    * Cloud Type 0
+    * Cloud Type 1
+    * Cloud Type 2
+    * Cloud Type 3
+    * Cloud Type 4
+    * Cloud Type 5
+    * Cloud Type 6
+    * Cloud Type 7
+    * Cloud Type 8
+    * Cloud Type 9
+    * Cloud Type 10
+    * Cloud Type 11
+    * Cloud Type 12
+    * Fill Flag 0
+    * Fill Flag 1
+    * Fill Flag 2
+    * Fill Flag 3
+    * Fill Flag 4
+    * Fill Flag 5
+    * Surface Albedo Units
+    * Version
+
+    The second item is a dataframe with the PSM3 timeseries data.
+
+    See Also
+    --------
+    pvlib.iotools.get_psm3
+
+    References
+    ----------
+    .. [1] `NREL National Solar Radiation Database (NSRDB)
+       <https://nsrdb.nrel.gov/>`_
+    .. [2] `Standard Time Series Data File Format
+       <https://rredc.nrel.gov/solar/old_data/nsrdb/2005-2012/wfcsv.pdf>`_
+    """
+    if hasattr(filename, 'readline'):
+        # if passed a file-like object, not our job to close it
+        close = False
+        fbuf = filename
+    else:
+        close = True
+        fbuf = open(filename, 'r')
```

Comment:
Is [duck typing](https://en.m.wikipedia.org/wiki/Duck_typing) possible here instead of `if-then`? I believe it would be more Pythonic:
```python
try:
    # always use binary, in case of Windows?
    fbuf = open(filename, 'rb')
except (OSError, IOError):
    pass
```

Also you can use `with` if you delegate passing the contents to a private function, _EG_:

```python
def _parse_psm3(fbuf):
    # The first 2 lines of the response are headers with metadata
    header_fields = fbuf.readline().split(',')

    # etc. all the stuff after try
    return contents

def read_psm3(fbuf):
    """..."""
    try:
        with = open(filename, 'rb') as fbuf:
            content = _parse_psm3(fbuf)
    except (OSError, IOError):
        content = _parse_psm3(fbuf)
    return content
```

Reply:
Agreed that the private `_parse_psm3` function is better.

I'll nitpick that using `hasattr(obj, 'method')` is an example of duck typing since it's asking "does `filename` `readline()` like a duck?", but I agree that trying to open the file would be more pythonic -- I'm just not sure that combining it with `with open...` is a good idea.  The try/except is wrapping the `_parse_psm3` call as well as the `open()` call when really we only want to catch errors from `open()` and I'm not sure it's necessarily possible to only catch errors from the `open` call.  Catching errors from both would cause the function to try to interpret a filename as a file-like object instead of raising a useful error.  For instance calling `open` on a `StringIO` raises a TypeError so we would have to trap `TypeError`, causing any TypeError raised in `_parse_psm3` to make the code try going down the other path and raise `AttributeError: 'str' object has no attribute 'readline'` as the primary error -- not great.  

I'm not opposed to using try/except for duck typing, but I'm not sure there's a way to safely combine it with a `with open(..) as fbuf:` call.  I don't like storing file handles outside of `with` and keeping track of whether or not to close the object, but I can't think of a better way to separate errors from `open()` and `_parse_psm3`:

```
def read_psm3(fbuf):
    """..."""
    try:
        # using rb causes error later on with fbuf.readline().split(',')
        fbuf = open(filename, 'r')
        close = True
    except TypeError:
        fbuf = filename
        close = False
    finally:
    
    try:
        content = _parse_psm3(fbuf)
    finally:
        if close:
            fbuf.close()
    return content
```

I know using try/except is generally looked on more favorably than attribute testing (EAFP vs LBYL), but to me it seems better to just ask the object "can I `readline` you?" than try to figure out what exceptions `open` might raise.  I'm not even sure that the above code is correct -- if `open` can raise TypeError in other situations besides when we give it a StringIO, we'd want to propagate that to the user.

In case you're interested, I got the `hasattr` pattern from [pandas's CSV writing code](https://github.com/pandas-dev/pandas/blob/master/pandas/io/formats/csvs.py#L177), though I think the additional complexity of that situation warrants the more flexible solution. 

Also RE: the file mode -- using 'rb' would require some other changes since you have to split a bytes object with a bytes object, not a str.  The tests pass on windows for me, so do we need to worry about it?

Reply:
Ok, I agree, "simple is better than complex, but complex is better than complicated." Looks good as is. I'll try to get through the rest later today. Thanks again!